tycho-gpg-plugin/src/main/java/org/eclipse/tycho/gpg/SignRepositoryArtifactsMojo.java

```diff
@@ -30,213 +26,269 @@
 public class SignRepositoryArtifactsMojo extends AbstractGpgMojoExtension {
+    enum PGPKeyBehavior {
+        skip, replace, merge
+    }
 
     /**
      * The maven project.
      */
     @Parameter(defaultValue = "${project}", readonly = true, required = true)
     protected MavenProject project;
 
+    /**
+     * The repository location.
+     */
     @Parameter(defaultValue = "${project.build.directory}/repository")
     private File repository;
 
     /**
-     * Configures to <code>true</code> to generate PGP signature only for artifacts that do
-     * <strong>not</strong> already contain signatures files from jarsigner.
+     * Configure to {@code true} to generate PGP signature only for artifacts that are
+     * <strong>not</strong> signed by a jarsigner.
      */
     @Parameter(defaultValue = "true")
     private boolean skipIfJarsigned;
 
     /**
-     * If set to {@code true} adds the public key belonging to the signature to the repository
+     * Configure to {@code true} to generate a PGP signature only for artifacts that do
+     * <strong>not</strong> contain a signature that's anchored in Java's trust store, i.e.,
+     * anchored in the JDK's cacerts. A JCA certificate, for example, is never anchored.
+     */
+    @Parameter(defaultValue = "true")
+    private boolean skipIfJarsignedAndAnchored;
+
+    /**
+     * Configure to {@code true} to generate a PGP signature for binary artifacts.
+     */
+    @Parameter(defaultValue = "true")
+    boolean skipBinaries;
+
+    /**
+     * Configure to {@code true} to add the public key of each signature to the repository's
      * metadata.
      */
     @Parameter(defaultValue = "true")
     private boolean addPublicKeyToRepo;
 
     /**
-     * If set to {@code true} adds the public key belonging to the signature to each signed
-     * artifact's metadata.
+     * Configure to {@code true} to add the public key of the signature to each signed artifact's
+     * metadata.
      */
     @Parameter(defaultValue = "true", alias = "addPublicKeysToArtifacts")
     private boolean addPublicKeyToArtifacts;
 
     /**
-     * Bundles that should be signed independently of other settings, e.g. {@link #skipIfJarsigned}.
+     * Configures how to generate PGP signatures for artifacts that already have one or more PGP
+     * signatures, {@code skip} to generate no new PGP signature, {@code replace} to replace the
+     * existing signature(s) with a new signature, and {@code merge} to add a new signature to any
+     * existing signature(s).
+     */
+    @Parameter(defaultValue = "skip")
+    private PGPKeyBehavior pgpKeyBehavior;
+
+    /**
+     * Configure to {@code false} to sign using Bouncy Castle libraries rather than launching
+     * {@code gpg} for that purpose. This is much faster and it runs in parallel, so is very much
+     * faster.
+     */
+    @Parameter(defaultValue = "true")
+    private boolean signWithGnuPG;
```

Comment:
This should become a String parameter (e.g. `signer`)

Then you can have
```
@Component(role = SignerFactory.class)
private Map<String, SignerFactory>  factories;
```

and have a one or more impls e.g.
```
@Component(role = SignerFactory.class, hint="my")
MySignerFactory implements SignerFactory {

... 

}
```
The even users can supply their own implementation if desired later on...

Reply:
Yes this is a bit bad (we probably ask maven-pgp-plugin to make this more open).

Maybe we still can use a string field with the values "bc" or "gpg" so if in the future we want to make it open for more implementations?

Reply:
Oh, that's a good idea!  Plan ahead for potential future flexibility.